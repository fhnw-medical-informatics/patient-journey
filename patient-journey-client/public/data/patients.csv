Day,Date
pid,date
2023-2-3,03.02.2023
2023-2-4,04.02.2023
2023-2-5,05.02.2023
2023-2-6,06.02.2023
2023-2-7,07.02.2023
2023-2-8,08.02.2023
2023-2-9,09.02.2023
2023-2-10,10.02.2023
2023-2-11,11.02.2023
2023-2-12,12.02.2023
2023-2-13,13.02.2023
2023-2-14,14.02.2023
2023-2-15,15.02.2023
2023-2-16,16.02.2023
2023-2-17,17.02.2023
2023-2-18,18.02.2023
2023-2-19,19.02.2023
2023-2-20,20.02.2023
2023-2-21,21.02.2023
2023-2-22,22.02.2023
2023-2-23,23.02.2023
2023-2-24,24.02.2023
2023-2-25,25.02.2023
2023-2-26,26.02.2023
2023-2-27,27.02.2023
2023-2-28,28.02.2023
2023-3-1,01.03.2023
2023-3-2,02.03.2023
2023-3-3,03.03.2023
2023-3-4,04.03.2023
2023-3-5,05.03.2023
2023-3-6,06.03.2023
2023-3-7,07.03.2023
2023-3-8,08.03.2023
2023-3-9,09.03.2023
2023-3-10,10.03.2023
2023-3-11,11.03.2023
2023-3-12,12.03.2023
2023-3-13,13.03.2023
2023-3-14,14.03.2023
2023-3-15,15.03.2023
2023-3-16,16.03.2023
2023-3-17,17.03.2023
2023-3-18,18.03.2023
2023-3-19,19.03.2023
2023-3-20,20.03.2023
2023-3-21,21.03.2023
2023-3-22,22.03.2023
2023-3-23,23.03.2023
2023-3-24,24.03.2023
2023-3-25,25.03.2023
2023-3-26,26.03.2023
2023-3-27,27.03.2023
2023-3-28,28.03.2023
2023-3-29,29.03.2023
2023-3-30,30.03.2023
2023-3-31,31.03.2023
2023-4-1,01.04.2023
2023-4-2,02.04.2023
2023-4-3,03.04.2023
2023-4-4,04.04.2023
2023-4-5,05.04.2023
2023-4-6,06.04.2023
2023-4-7,07.04.2023
2023-4-8,08.04.2023
2023-4-9,09.04.2023
2023-4-10,10.04.2023
2023-4-11,11.04.2023
2023-4-12,12.04.2023
2023-4-13,13.04.2023
2023-4-14,14.04.2023
2023-4-15,15.04.2023
2023-4-16,16.04.2023
2023-4-17,17.04.2023
2023-4-18,18.04.2023
2023-4-19,19.04.2023
2023-4-20,20.04.2023
2023-4-21,21.04.2023
2023-4-22,22.04.2023
2023-4-23,23.04.2023
2023-4-24,24.04.2023
2023-4-25,25.04.2023
2023-4-26,26.04.2023
2023-4-27,27.04.2023
2023-4-28,28.04.2023
2023-4-29,29.04.2023
2023-4-30,30.04.2023
2023-5-1,01.05.2023
2023-5-2,02.05.2023
2023-5-3,03.05.2023
2023-5-4,04.05.2023
2023-5-5,05.05.2023
2023-5-6,06.05.2023
2023-5-7,07.05.2023
2023-5-8,08.05.2023
2023-5-9,09.05.2023
2023-5-10,10.05.2023
2023-5-11,11.05.2023
2023-5-12,12.05.2023
2023-5-13,13.05.2023
2023-5-14,14.05.2023
2023-5-15,15.05.2023
2023-5-16,16.05.2023
2023-5-17,17.05.2023
2023-5-18,18.05.2023
2023-5-19,19.05.2023
2023-5-20,20.05.2023
2023-5-21,21.05.2023
2023-5-22,22.05.2023
2023-5-23,23.05.2023
2023-5-24,24.05.2023
2023-5-25,25.05.2023
2023-5-26,26.05.2023
2023-5-27,27.05.2023
2023-5-28,28.05.2023
2023-5-29,29.05.2023
2023-5-30,30.05.2023
2023-5-31,31.05.2023
2023-6-1,01.06.2023
