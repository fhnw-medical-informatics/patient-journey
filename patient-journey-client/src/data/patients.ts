import { DataColumn, GenericColumnType } from './columns'
import { ParseResult } from 'papaparse'
import { DataEntity, Entity, EntityId, EntityIdNone } from './entities'
import { noOp } from '../utils'

export const PATIENT_ID_COLUMN_TYPE = 'pid'

export type PatientDataColumnType = typeof PATIENT_ID_COLUMN_TYPE | GenericColumnType
export type PatientDataColumn = DataColumn<PatientDataColumnType>

export interface Patient extends Entity {
  readonly type: 'patients'
  readonly pid: PatientId
}

enum PatientIdBrand {}

export type PatientId = PatientIdBrand & string
export const PatientIdNone = 'n/a' as PatientId

export interface PatientData extends DataEntity<Patient, PatientDataColumn> {}

export const EMPTY_PATIENT_DATA: PatientData = {
  type: 'patients',
  columns: [],
  allEntities: [],
  selectedEntity: EntityIdNone,
  hoveredEntity: EntityIdNone,
}

export const createPatientData = (
  result: ParseResult<string[]>,
  onWarning: (message: string) => void = noOp,
  onError: (message: string) => void = noOp
): PatientData => {
  const HEADER_ROW_COUNT = 2
  if (result.data.length < HEADER_ROW_COUNT) {
    onError('Patient data table must contain two header rows (column names, column types).')
    return EMPTY_PATIENT_DATA
  } else {
    const columnNames = result.data[0]
    const columnTypes = result.data[1].map((v) => v.toLowerCase())
    const idColumnIndex = columnTypes.indexOf(PATIENT_ID_COLUMN_TYPE)
    const isMissingIdColumn = idColumnIndex < 0

    if (isMissingIdColumn) {
      onWarning(
        `No '${PATIENT_ID_COLUMN_TYPE}' column type found in patient data table. Using row index to identify patients.`
      )
    }

    const columns = columnNames.map<PatientDataColumn>((name, index) => ({
      name,
      type: columnTypes[index] as PatientDataColumnType,
      index,
    }))

    return {
      ...EMPTY_PATIENT_DATA,
      columns,
      allEntities: result.data.slice(HEADER_ROW_COUNT).map((row: string[], index) => {
        const id = isMissingIdColumn ? String(index) : row[idColumnIndex]
        return {
          uid: id as EntityId,
          type: 'patients',
          pid: id as PatientId,
          values: row,
        }
      }),
    }
  }
}
