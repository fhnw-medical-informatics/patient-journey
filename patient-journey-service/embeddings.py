import os

import json

from dotenv import load_dotenv
from openai import OpenAI
from chunks import create_patient_journeys_chunks

# Load the environment variables from the .env file
load_dotenv(".env.local")

# Sources:
# https://platform.openai.com/docs/guides/embeddings/embedding-models
# https://openai.com/pricing
EMBEDDINGS_API_COSTS_PER_1KTOKENS = 0.0001
MODEL = "text-embedding-ada-002"
TOKENS_PER_CHUNK = 6000 # text-embedding-ada-002 has a limit of 8191 tokens per request

openaiAPI = OpenAI(
    api_key=os.environ.get("OPENAI_API_KEY"),
    max_retries=5
)

def create_embeddings_for_chunk(chunk: list[str], model: str = "text-embedding-ada-002") -> list[list[float]]:
    """
    Generate embeddings for a given chunk of text using the specified model.

    :param chunk: A list of strings for which to generate embeddings.
    :param model: The model to use for generating embeddings. Defaults to "text-embedding-ada-002".
    :return: A list of embeddings, where each embedding corresponds to an item in the input chunk.
    :raises Exception: If the embeddings cannot be generated or the response is invalid.
    """
    if not chunk:
        return []

    try:
        embeddings_response = openaiAPI.embeddings.create(model=model, input=chunk)
    except Exception as e:
        raise Exception(f"Failed to generate embeddings: {e}")

    if not embeddings_response or len(embeddings_response.data) != len(chunk):
        raise Exception("Invalid response from embeddings API.")

    return [embedding.embedding for embedding in embeddings_response.data]

def resumable_create_embeddings(patient_journeys: list[str], journeys_hash: str) -> list[list[float]]:
    """
    Generate embeddings for a list of patient journeys.
    But if an exception occurs, the function will store the partial results in a file and raise an exception.
    When the function is run again with the same input, it will load the partial results from the file and continue where it left off.

    Since the embeddings are created in the same order as the partient_journeys list, when a partial result is loaded from the file,
    the function will skip the patient journeys for which embeddings were already generated and append the new embeddings to the partial results.

    If another error happens, the function will do the same again, storing the new partial results in the file. Until all embeddings are generated.

    In order to identify the partial results file, the function will use the hash of the input list of patient journeys.

    Once all embeddings are generated, the function will delete the partial results file.

    This function uses the create_embeddings function to generate the embeddings.

    :param patient_journeys: A list of patient journey strings.
    :return: A list of embeddings for each patient journey.
    :raises Exception: If not all embeddings are generated.
    """
    # Calculate a hash of the patient_journeys list to use as a filename
    partial_results_file = f"partial_embeddings_{journeys_hash}.json"

    resumable_data = get_remaining_journeys_and_cached_embeddings_from_file(patient_journeys, partial_results_file)

    partial_embeddings = resumable_data['partial_embeddings']
    remaining_patient_journeys = resumable_data['remaining_patient_journeys']

    try:
        # Generate embeddings for the remaining patient journeys
        new_embeddings = create_embeddings(remaining_patient_journeys)
        partial_embeddings.extend(new_embeddings)

        if len(partial_embeddings) != len(patient_journeys):
            raise Exception("Everything was successful, but not all embeddings were generated.", "Attaching Partial Embeddings", partial_embeddings)
        
    except Exception as e:
        # Merge the created embeddings from the exception with the existing partial_embeddings
        if hasattr(e, 'args') and len(e.args) > 2 and e.args[1] == "Attaching Partial Embeddings" and isinstance(e.args[2], list):
            partial_embeddings.extend(e.args[2])

            # Save the merged embeddings to the file
            save_partial_embeddings_to_file(partial_embeddings, partial_results_file)

            print(f"⚠️ An exception occurred during embeddings generation: {e.args[0]}, the attempt is resumable with the same hash ({journeys_hash}).")
            raise Exception(f"An exception occurred during embeddings generation: {e.args[0]}", "Partial Embeddings Saved")
        else:
            print(f"🚨 An exception occurred during embeddings generation: {e}, the attempt is NOT resumable.")
            raise Exception(f"An exception occurred during embeddings generation: {e}")

    # If we've generated all embeddings, delete the partial results file    
    cleanup_partial_embeddings_file(partial_results_file)

    return partial_embeddings

def create_embeddings(patient_journeys: list[str]) -> list[list[float]]:
    """Generate embeddings for a list of patient journeys.

    :param patient_journeys: A list of patient journey strings.
    :return: A list of embeddings for each patient journey.
    :raises Exception: If not all embeddings are generated.
    """
    chunks_info = create_patient_journeys_chunks(patient_journeys, MODEL, TOKENS_PER_CHUNK)
    patient_journey_chunks = chunks_info['patient_journey_chunks']
    total_tokens = chunks_info['total_nr_of_tokens']
    num_chunks = len(patient_journey_chunks)

    print_embedding_generation_info(len(patient_journeys), num_chunks, total_tokens)

    partial_embeddings = []
    try:
        for i, chunk in enumerate(patient_journey_chunks):
            display_progress(i, num_chunks)
            partial_embeddings.extend(generate_chunk_embeddings(chunk, i, num_chunks))
        
        validate_embeddings_count(partial_embeddings, patient_journeys)
    except Exception as e:
        raise Exception(f"An exception occurred during embeddings generation: {e}", "Attaching Partial Embeddings", partial_embeddings)
    
    return partial_embeddings

# Recover potentially cached embeddings form a previous run
def get_remaining_journeys_and_cached_embeddings_from_file(patient_journeys: list[str], partial_results_file: str) -> dict:
    # Try to load partial results if they exist
    partial_embeddings = []
    
    if os.path.exists(partial_results_file):
        with open(partial_results_file, 'r') as file:
            partial_embeddings = json.load(file)
            print(f"📂 Resuming from previously created embeddings ({len(partial_embeddings)}/{len(patient_journeys)}) in file: {partial_results_file}")

    # Calculate the starting index based on the number of embeddings already generated
    start_index = len(partial_embeddings)

    # Only process the remaining patient journeys
    return {
        "partial_embeddings": partial_embeddings,
        "remaining_patient_journeys": patient_journeys[start_index:]
    }

def save_partial_embeddings_to_file(partial_embeddings: list[list[float]], partial_results_file: str):
    # Save the merged embeddings to the file
    with open(partial_results_file, 'w') as file:
        json.dump(partial_embeddings, file)

def cleanup_partial_embeddings_file(partial_results_file: str): 
    if os.path.exists(partial_results_file):
        os.remove(partial_results_file)
        print(f"🧹 Cleaned up partial embeddings file: {partial_results_file}")

def print_embedding_generation_info(num_journeys: int, num_chunks: int, total_tokens: int):
    """Prints information about the embedding generation process."""
    print(f"🧩 Generating embeddings using the `{MODEL}` model.")
    print(f"Patient Journeys: {num_journeys}")
    print(f"Chunks: {num_chunks}")
    print(f"Tokens: {total_tokens}")
    estimated_cost = (total_tokens / 1000) * EMBEDDINGS_API_COSTS_PER_1KTOKENS
    print(f"====================\n💸 Estimated costs: ${estimated_cost:.4f}\n====================\n")

def display_progress(current_chunk: int, total_chunks: int):
    """Displays the progress of the embedding generation."""
    progress = (current_chunk + 1) / total_chunks
    progress_bar = ('#' * int(progress * 20)).ljust(20)
    print(f"\r[{progress_bar}] Chunk {current_chunk + 1}/{total_chunks} being sent to openai embeddings api.", end='')

def generate_chunk_embeddings(chunk: list[str], chunk_index: int, total_chunks: int) -> list[list[float]]:
    """Generates embeddings for a single chunk and handles exceptions."""
    try:
        return create_embeddings_for_chunk(chunk, MODEL)
    except Exception as e:
        error_message = f"\nAn error occurred while generating embeddings for chunk {chunk_index + 1} of {total_chunks}: {e}"
        print(error_message)
        raise Exception(error_message)

def validate_embeddings_count(embeddings: list[list[float]], patient_journeys: list[str]):
    """Validates that the number of generated embeddings matches the number of patient journeys."""
    if len(embeddings) != len(patient_journeys):
        error_message = f"Number of embeddings generated ({len(embeddings)}) does not match the number of patient journeys ({len(patient_journeys)})."
        print(error_message)
        raise Exception("Not all embeddings were generated.")
    print(f"✅ Finished generating {len(embeddings)} embeddings for {len(patient_journeys)} patient journeys.\n")


