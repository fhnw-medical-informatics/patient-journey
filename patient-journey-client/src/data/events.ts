import { DataColumn, GENERIC_COLUMN_TYPES } from './columns'
import { PATIENT_ID_COLUMN_TYPE, PatientId } from './patients'
import { ParseResult } from 'papaparse'
import { DataEntity, Entity, EntityId } from './entities'
import { noOp } from '../utils'

export const EVENT_ID_COLUMN_TYPE = 'eid' as const

export const EVENT_DATA_COLUMN_TYPES = [...GENERIC_COLUMN_TYPES, EVENT_ID_COLUMN_TYPE, PATIENT_ID_COLUMN_TYPE] as const
export type EventDataColumnType = typeof EVENT_DATA_COLUMN_TYPES[number]

export type EventDataColumn = DataColumn<EventDataColumnType>

enum EventIdBrand {}

export type EventId = EventIdBrand & string

export interface PatientJourneyEvent extends Entity {
  readonly eid: EventId
  readonly pid: PatientId
}

export interface EventData extends DataEntity<PatientJourneyEvent, EventDataColumn> {}

export const EMPTY_EVENT_DATA: EventData = {
  columns: [],
  allEntities: [],
}

export const createEventData = (
  result: ParseResult<string[]>,
  headerRowCount: number,
  onWarning: (message: string) => void = noOp
): EventData => {
  const columnNames = result.data[0]
  const columnTypes = result.data[1].map((v) => v.toLowerCase())
  const eventIdColumnIndex = columnTypes.indexOf(EVENT_ID_COLUMN_TYPE)
  const isMissingEventIdColumn = eventIdColumnIndex < 0
  const patientIdColumnIndex = columnTypes.indexOf(PATIENT_ID_COLUMN_TYPE)
  const isMissingPatientIdColumn = patientIdColumnIndex < 0

  if (isMissingPatientIdColumn) {
    throw new Error(`No '${PATIENT_ID_COLUMN_TYPE}' column type found in event data table.`)
  }

  if (isMissingEventIdColumn) {
    onWarning(`No '${EVENT_ID_COLUMN_TYPE}' column type found in event data table. Using row index to identify events.`)
  }

  columnTypes.forEach((type) => {
    if (!isEventDataColumnType(type)) {
      onWarning(`Invalid column type '${type}' found in event data table. Falling back to 'string'.`)
    }
  })

  const columns = columnNames.map<EventDataColumn>((name, index) => ({
    name,
    type: columnTypes[index] as EventDataColumnType,
    index,
  }))
  return {
    ...EMPTY_EVENT_DATA,
    columns,
    allEntities: result.data.slice(headerRowCount).map((row: string[], index) => {
      const id = isMissingEventIdColumn ? String(index) : row[eventIdColumnIndex]
      return {
        uid: id as EntityId,
        eid: id as EventId,
        pid: row[patientIdColumnIndex] as PatientId,
        values: row,
      }
    }),
  }
}

export const isEventDataColumnType = (columnType: string): boolean =>
  EVENT_DATA_COLUMN_TYPES.includes(columnType as EventDataColumnType)
