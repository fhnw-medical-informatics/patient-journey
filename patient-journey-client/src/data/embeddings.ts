import { encode } from '@nem035/gpt-3-encoder'
import TSNE from 'tsne-js'
import kMeans from 'kmeans-js'

import { EventData } from './events'
import { PatientData } from './patients'
import { openaiAPI } from '../utils/openai'
import { sha256 } from '../utils'
import { EMBEDDINGS_DATA_FILE_URL } from './constants'

export const EMBEDDINGS_API_COSTS_PER_1KTOKENS = 0.0004
export const TOKENS_PER_CHUNK = 6000

export type Embeddings = Record<string, ReadonlyArray<number>>

export type EmbeddingsFile = { patientDataHash: string; embeddings: Embeddings }

export type EmbeddingsStateLoadingPending = Readonly<{
  type: 'loading-pending'
}>

export type EmbeddingsStateLoadingInProgress = Readonly<{
  type: 'loading-in-progress'
}>

export type EmbeddingsStateLoadingFailed = Readonly<{
  type: 'loading-failed'
  errorMessage: string
}>

export interface LoadedEmbeddings {
  embeddings: Embeddings
}

export interface ReducedEmbeddings {
  reducedEmbeddings: Embeddings
}

export interface Clusters {
  clusters: Record<string, number>
}

export interface LoadedPromptEmbeddings {
  embedding: number[]
}

export type EmbeddingsStateLoadingComplete = Readonly<{
  type: 'loading-complete'
}> &
  LoadedEmbeddings &
  ReducedEmbeddings &
  Clusters

export type PromptEmbeddingsStateLoadingComplete = Readonly<{
  type: 'loading-complete'
}> &
  LoadedPromptEmbeddings

export type EmbeddingsData = {
  readonly patientDataEmbeddings:
    | EmbeddingsStateLoadingPending
    | EmbeddingsStateLoadingInProgress
    | EmbeddingsStateLoadingFailed
    | EmbeddingsStateLoadingComplete
  readonly promptEmbeddings:
    | EmbeddingsStateLoadingPending
    | EmbeddingsStateLoadingInProgress
    | EmbeddingsStateLoadingFailed
    | PromptEmbeddingsStateLoadingComplete
}

export const preparePatientJourneys = (patientData: PatientData, eventData: EventData): Array<string> =>
  patientData.allEntities.map((patient) => {
    const events = eventData.allEntities.filter((event) => event.pid === patient.pid)

    return `
    Patient information:
      ${patientData.columns.map((column) => `${column.name}: ${patient.values[column.index]}`).join('\n')}

    The patients' journey through the hospital:
      ${events
        .map(
          (event, idx) =>
            `Event ${idx + 1}: ${eventData.columns
              .map((column) => `${column.name}: ${event.values[column.index]}`)
              .join(', ')}`
        )
        .join('\n')}
   `
  })

function delay(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms))
}

export async function retryOpenaiAPI(maxRetries: number, inputChunk: Array<string>) {
  const baseWaitTime = 30000 / 2 ** (maxRetries - 1)

  for (let i = 0; i < maxRetries; i++) {
    try {
      const response = await openaiAPI.createEmbedding({
        model: 'text-embedding-ada-002',
        input: inputChunk,
      })

      return response
    } catch (error) {
      console.error(`Error in retry ${i + 1}:`, error)

      if (i === maxRetries - 1) {
        throw error
      }

      const waitTime = Math.random() * 2 ** i * baseWaitTime
      await delay(waitTime)
    }
  }
}

/**
 * Create chunks of patient journeys to be sent to the openai embeddings api.
 * Each chunk is an array of patient journeys (strings) that are in total less
 * than or equal to maxTokensPerChunk.
 *
 * The openai embeddings api has a limit of 8191 tokens per request. The number of
 * tokens for a string can be calculated with the gpt-3-encoder package.
 *
 * @param journeys
 * @param maxTokensPerChunk
 */
export const createPatientJourneysChunks = (
  journeys: Array<string>,
  maxTokensPerChunk: number
): { totalNrOfTokens: number; patientJourneyChunks: Array<Array<string>> } => {
  const chunks: Array<Array<string>> = []
  let totalNrOfTokens = 0
  let currentChunk: string[] = []
  let currentChunkTokens = 0

  journeys.forEach((journey, idx) => {
    let journeyString = journey
    let journeyTokens = encode(journeyString).length

    // If a single journey is longer than the max number of tokens per chunk, shrink it until it fits
    if (journeyTokens > maxTokensPerChunk) {
      console.warn(
        `Patient journey ${idx} is longer than the max number of tokens per chunk. It has ${journeyTokens} tokens.`
      )

      while (journeyTokens > maxTokensPerChunk) {
        journeyString = journeyString.slice(0, -1000)
        journeyTokens = encode(journeyString).length
      }

      console.warn(
        `Patient journey ${idx} was longer than the max number of tokens per chunk. It was shrunk to fit. Now it has ${journeyTokens} tokens.`
      )
    }

    if (currentChunkTokens + journeyTokens > maxTokensPerChunk) {
      chunks.push(currentChunk)
      currentChunk = []
      currentChunkTokens = 0
    }

    currentChunk.push(journeyString)
    currentChunkTokens += journeyTokens
    totalNrOfTokens += journeyTokens
  })

  if (currentChunkTokens > 0) {
    chunks.push(currentChunk)
  }

  return {
    totalNrOfTokens,
    patientJourneyChunks: chunks,
  }
}

export const loadEmbeddings = async (patientData: PatientData, eventData: EventData): Promise<EmbeddingsData> => {
  const patientDataHash = await sha256(JSON.stringify(patientData.allEntities))

  // Check if embeddings are already cached in local storage
  const cachedEmbeddingsFile = await loadEmbeddingsFileFromUrl(EMBEDDINGS_DATA_FILE_URL)

  if (cachedEmbeddingsFile && cachedEmbeddingsFile.patientDataHash === patientDataHash) {
    console.log('Embeddings loaded from cache in local storage')

    const embeddings = cachedEmbeddingsFile.embeddings

    console.log('Embeddings', embeddings)

    const reducedEmbeddings = reduceEmbeddings(embeddings)
    const clusters = clusterEmbeddings(reducedEmbeddings, 3)

    return Promise.resolve({
      patientDataEmbeddings: {
        type: 'loading-complete',
        embeddings,
        reducedEmbeddings,
        clusters,
      },
      promptEmbeddings: {
        type: 'loading-pending',
      },
    })
  } else {
    if (cachedEmbeddingsFile && cachedEmbeddingsFile.patientDataHash !== patientDataHash) {
      console.log('Cached embeddings are for a different patient data set, loading new embeddings from API')
    }

    const patientJourneys = preparePatientJourneys(patientData, eventData)

    const { totalNrOfTokens, patientJourneyChunks } = createPatientJourneysChunks(patientJourneys, TOKENS_PER_CHUNK) // 8191 is the max number of tokens per request (but encode library does not seem to be exact)

    console.log('Total number of tokens:', totalNrOfTokens)
    console.log('Number of chunks:', patientJourneyChunks.length)
    console.log('Number of patient Journeys', patientJourneys.length)

    console.log('Estimated cost:', (totalNrOfTokens / 1000) * EMBEDDINGS_API_COSTS_PER_1KTOKENS)

    const patientJourneyEmbeddings: Array<Array<number>> = []

    for (const [chunkIdx, chunk] of patientJourneyChunks.entries()) {
      console.log(`Sending chunk ${chunkIdx + 1} of ${patientJourneyChunks.length} to openai embeddings api`)

      try {
        const chunkEmbeddings = await retryOpenaiAPI(3, chunk)

        if (chunkEmbeddings && chunkEmbeddings.data.data.length > 0) {
          chunkEmbeddings.data.data.forEach((embedding) => {
            patientJourneyEmbeddings.push(embedding.embedding)
          })
        } else {
          console.error(`Failed to get embeddings for chunk ${chunkIdx + 1}`)

          return {
            patientDataEmbeddings: {
              type: 'loading-failed',
              errorMessage: 'Something went wrong while loading the embeddings',
            },
            promptEmbeddings: {
              type: 'loading-pending',
            },
          }
        }
      } catch (error) {
        console.error(`Failed to get embeddings for chunk ${chunkIdx + 1} after 3 retries`, error)

        return {
          patientDataEmbeddings: {
            type: 'loading-failed',
            errorMessage: 'Something went wrong while loading the embeddings',
          },
          promptEmbeddings: {
            type: 'loading-pending',
          },
        }
      }
    }

    console.log('Number of patient journey embeddings:', patientJourneyEmbeddings.length)

    if (patientJourneyEmbeddings.length !== patientData.allEntities.length) {
      console.error(
        'Number of patient journey embeddings does not match number of patient journeys. Something went wrong.'
      )

      return {
        patientDataEmbeddings: {
          type: 'loading-failed',
          errorMessage: 'Something went wrong while loading the embeddings',
        },
        promptEmbeddings: {
          type: 'loading-pending',
        },
      }
    }

    const patientDataEmbeddings: Embeddings = {}

    patientData.allEntities.forEach((patient, idx) => {
      patientDataEmbeddings[patient.pid] = patientJourneyEmbeddings[idx]
    })

    // Download embeddings as json file and prompt user to save it
    const blob = new Blob([JSON.stringify({ patientDataHash, embeddings: patientDataEmbeddings })], {
      type: 'application/json',
    })
    const url = URL.createObjectURL(blob)
    console.log('Download embeddings as json file:', url)
    // <a href="blob:http://127.0.0.1:3000/77094bbb-5b03-4391-92dc-42796fd8ba3d" download="embeddings.json">Download</a>

    console.log('Embeddings cached successfully')

    console.log('Embeddings', patientDataEmbeddings)

    const reducedEmbeddings = reduceEmbeddings(patientDataEmbeddings)
    const clusters = clusterEmbeddings(reducedEmbeddings, 3)

    return Promise.resolve({
      patientDataEmbeddings: {
        type: 'loading-complete',
        embeddings: patientDataEmbeddings,
        reducedEmbeddings,
        clusters,
      },
      promptEmbeddings: {
        type: 'loading-pending',
      },
    })
  }
}

const loadEmbeddingsFileFromUrl = async (url: string): Promise<EmbeddingsFile | undefined> => {
  const response = await fetch(url)

  if (response.ok) {
    const json = await response.json()

    return json
  } else {
    return undefined
  }
}

/**
 * Generate reduced 2d embeddings using t-SNE.
 *
 * @param embeddings High dimensional embeddings
 */
const reduceEmbeddings = (embeddings: Embeddings): Embeddings => {
  const tsne = new TSNE({
    dim: 2,
    perplexity: 30.0,
    earlyExaggeration: 4.0,
    learningRate: 100.0,
    nIter: 500,
    metric: 'euclidean',
  })

  const data = Object.values(embeddings)

  tsne.init({
    data,
    type: 'dense',
  })

  tsne.on('progressIter', (iter: number) => {
    console.log('Iteration', iter)
  })

  tsne.run()

  // for (let k = 0; k < 500; k++) {
  //   tsne.rerun()
  // }

  // Is the rerun necessary?
  // tsne.rerun()

  const Y = tsne.getOutput()

  const reducedEmbeddings: Embeddings = {}

  Object.keys(embeddings).forEach((pid, idx) => {
    reducedEmbeddings[pid] = Y[idx]
  })

  console.log('Reduced embeddings', reducedEmbeddings)

  return reducedEmbeddings
}

/**
 * Perform K-means clustering on the reduced embeddings
 * @param reducedEmbeddings Embeddings reduced to 2 dimensions
 * @param k Number of clusters
 * @returns Clustered embeddings
 */
const clusterEmbeddings = (reducedEmbeddings: Embeddings, k: number): Record<string, number> => {
  const data = Object.values(reducedEmbeddings)

  const km = new kMeans({
    k,
  })

  km.cluster(data)

  while (km.step()) {
    km.findClosestCentroids()
    km.moveCentroids()

    if (km.hasConverged()) break
  }

  console.log('Cluster centroids', km.centroids)

  // An array of arrays containing the indices of the data points that belong to each cluster
  console.log('Cluster assignments', km.clusters)

  const clusteredEmbeddings: Record<string, number> = {}

  const patientIds = Object.keys(reducedEmbeddings)

  for (let clusterIdx = 0; clusterIdx < km.clusters.length; clusterIdx++) {
    const cluster = km.clusters[clusterIdx]

    for (const dataPointIdx of cluster) {
      clusteredEmbeddings[patientIds[dataPointIdx]] = clusterIdx
    }
  }

  console.log('Clustered embeddings', clusteredEmbeddings)

  return clusteredEmbeddings
}
