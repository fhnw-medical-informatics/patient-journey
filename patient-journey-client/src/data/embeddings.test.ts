import { preparePatientJourneys } from './embeddings'
import { EventData, EventId } from './events'
import { PatientData, PatientId } from './patients'

describe('preparePatientJourneys', () => {
  // match snapshot
  it('should return a patient journey as a string', () => {
    const patientData: PatientData = {
      allEntities: [
        {
          pid: '1' as PatientId,
          uid: '1' as PatientId,
          values: ['Patient 1', 'Male'],
        },
        {
          pid: '2' as PatientId,
          uid: '2' as PatientId,
          values: ['Patient 2', 'Female'],
        },
      ],
      columns: [
        {
          name: 'Patient',
          type: 'string',
          index: 0,
        },
        {
          name: 'Gender',
          type: 'string',
          index: 1,
        },
      ],
    }

    const eventData: EventData = {
      allEntities: [
        {
          eid: '1' as EventId,
          pid: '1' as PatientId,
          uid: '1' as EventId,
          values: ['Event 1', '2019-01-01'],
        },
        {
          eid: '2' as EventId,
          pid: '1' as PatientId,
          uid: '2' as EventId,
          values: ['Event 2', '2019-01-02'],
        },
        {
          eid: '3' as EventId,
          pid: '2' as PatientId,
          uid: '3' as EventId,
          values: ['Event 3', '2019-01-03'],
        },
      ],
      columns: [
        {
          name: 'Event',
          type: 'string',
          index: 0,
        },
        {
          name: 'Date',
          type: 'date',
          index: 1,
        },
      ],
    }

    const patientJourneys = preparePatientJourneys(patientData, eventData)
    expect(patientJourneys).toMatchSnapshot()
  })
})
