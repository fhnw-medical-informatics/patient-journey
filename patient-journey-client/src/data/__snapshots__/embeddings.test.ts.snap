// Vitest Snapshot v1

exports[`preparePatientJourneys > should return a patient journey as a string 1`] = `
[
  "
  Patient information:
    Patient: Patient 1
    Gender: Male

  The patients' journey through the hospital:
    Event 0: 
      Event: Event 1
      Date: 2019-01-01
    Event 1: 
      Event: Event 2
      Date: 2019-01-02
  ",
  "
  Patient information:
    Patient: Patient 2
    Gender: Female

  The patients' journey through the hospital:
    Event 0: 
      Event: Event 3
      Date: 2019-01-03
  ",
]
`;
